src/clusterfuzz/_internal/system/archive.py

```diff
@@ -33,116 +39,356 @@
+
+  def extract(self,
+              member: str,
+              path: Union[str, os.PathLike],
+              trusted: bool = False):
+    # If the output directory is a symlink, get its actual path since we will be
+    # doing directory traversal checks later when unpacking the archive.
+    output_directory = os.path.realpath(path)
+
+    # If the archive is not trusted, do file path checks to make
+    # sure this archive is safe and is not attempting to do path
+    # traversals.
+    if not trusted:
+      if (_is_attempting_path_traversal(self._zip_archive.filename,
+                                        output_directory, member)):
+        return None
 
-def extracted_size(archive_path, archive_obj=None, file_match_callback=None):
-  """Return the total extracted size of the archive."""
-  return sum(
-      f.size for f in iterator(
-          archive_path, archive_obj, file_match_callback, should_extract=False))
+    try:
+      extracted_path = self._zip_archive.extract(member, output_directory)
+
+      # Preserve permissions for regular files. 640 is the default
+      # permission for extract. If we need execute permission, we need
+      # to chmod it explicitly. Also, get rid of suid bit for security
+      # reasons.
+      external_attr = self._zip_archive.getinfo(member).external_attr >> 16
+      if oct(external_attr).startswith(FILE_ATTRIBUTE):
+        old_mode = external_attr & 0o7777
+        new_mode = external_attr & 0o777
+        new_mode |= 0o440
+        needs_execute_permission = external_attr & 100
+
+        if new_mode != old_mode or needs_execute_permission:
+          # Default extract condition is 640 which is safe.
+          # |new_mode| might have read+write+execute bit for
+          # others, so remove those.
+          new_mode &= 0o770
+
+          os.chmod(extracted_path, new_mode)
+
+      # Create symlink if needed (only on unix platforms).
+      if (trusted and hasattr(os, 'symlink') and
+          oct(external_attr).startswith(SYMLINK_ATTRIBUTE)):
+        symlink_source = self._zip_archive.read(member)
+        if os.path.exists(extracted_path):
+          os.remove(extracted_path)
+        os.symlink(symlink_source, extracted_path)
+
+      return extracted_path
+    except Exception as e:
+      # In case of errors, we try to extract whatever we can without errors.
+      logs.log_warn('An error occured while extracting %s from the archive: %s.'
+                    % (member, repr(e)))
+      return None
+
+  def extract_all(self,
+                  path: Union[str, os.PathLike],
+                  members: Optional[List[str]] = None,
+                  trusted: bool = False) -> None:
+    to_extract = members if members is not None else self._zip_archive.namelist(
+    )
+    for member in to_extract:
+      self.extract(member=member, path=path, trusted=trusted)
+
+
+class ArchiveError(Exception):
+  """ArchiveError"""
+
+
+# pylint: disable=redefined-builtin
+def open(archive_path: str,
```

Comment:
It's weird that we are returning an archive-file but calling it reader everywhere. I suppose that is because therew will be a conflict because this module is named "archive" but you can do `import ... as ...` or we can rename this to archivelib/archive_utils (but not now)